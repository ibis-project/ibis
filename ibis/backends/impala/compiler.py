from __future__ import annotations

import ibis.expr.operations as ops
from ibis.backends.base.sql.compiler import Compiler, ExprTranslator, TableSetFormatter
from ibis.backends.base.sql.registry import binary_infix_ops, operation_registry, unary


class ImpalaTableSetFormatter(TableSetFormatter):
    def _get_join_type(self, op):
        jname = self._join_names[type(op)]

        # Impala requires this
        if not op.predicates:
            jname = self._join_names[ops.CrossJoin]

        return jname

    def _format_in_memory_table(self, op):
        if op.data:
            return super()._format_in_memory_table(op)

        schema = op.schema
        names = schema.names
        types = schema.types
        rows = [
            f"{self._translate(ops.Cast(ops.Literal(None, dtype=dtype), to=dtype))} AS {name}"
            for name, dtype in zip(map(self._quote_identifier, names), types)
        ]
        return f"(SELECT * FROM (SELECT {', '.join(rows)}) AS _ LIMIT 0)"


class ImpalaExprTranslator(ExprTranslator):
    _registry = {**operation_registry, **binary_infix_ops, ops.Hash: unary("fnv_hash")}
    _forbids_frame_clause = (
        *ExprTranslator._forbids_frame_clause,
        ops.Lag,
        ops.Lead,
        ops.FirstValue,
        ops.LastValue,
    )
    _unsupported_reductions = (
        ops.ApproxMedian,
        ops.ApproxCountDistinct,
        ops.GroupConcat,
    )
    _dialect_name = "hive"
    _quote_identifiers = True


rewrites = ImpalaExprTranslator.rewrites


@rewrites(ops.FloorDivide)
def _floor_divide(op):
    return ops.Floor(ops.Divide(op.left, op.right))


class ImpalaCompiler(Compiler):
    translator_class = ImpalaExprTranslator
    table_set_formatter_class = ImpalaTableSetFormatter
