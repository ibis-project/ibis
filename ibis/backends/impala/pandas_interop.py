from __future__ import annotations

# Copyright 2014 Cloudera Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import contextlib
import csv
import os
import tempfile
from posixpath import join as pjoin

import ibis.common.exceptions as com
import ibis.expr.schema as sch
from ibis import util
from ibis.config import options


class DataFrameWriter:
    """Interface class for writing pandas objects to Impala tables.

    Notes
    -----
    Class takes ownership of any temporary data written to HDFS
    """

    def __init__(self, client, df):
        self.client = client
        self.df = df
        self.temp_hdfs_dirs = set()

    def write_temp_csv(self):
        temp_hdfs_dir = pjoin(options.impala.temp_hdfs_path, f"pandas_{util.guid()}")
        self.client.hdfs.mkdir(temp_hdfs_dir)

        # Keep track of the temporary HDFS file
        self.temp_hdfs_dirs.add(temp_hdfs_dir)

        # Write the file to HDFS
        hdfs_path = pjoin(temp_hdfs_dir, "0.csv")

        self.write_csv(hdfs_path)

        return temp_hdfs_dir

    def write_csv(self, path):
        # Use a temporary dir instead of a temporary file
        # to provide Windows support and avoid #2267
        # https://github.com/ibis-project/ibis/issues/2267
        with tempfile.TemporaryDirectory() as f:
            # Write the DataFrame to the temporary file path
            tmp_file_path = os.path.join(f, "impala_temp_file.csv")
            if options.verbose:
                util.log(f"Writing DataFrame to temporary directory {tmp_file_path}")

            self.df.to_csv(
                tmp_file_path,
                header=False,
                index=False,
                sep=",",
                quoting=csv.QUOTE_NONE,
                escapechar="\\",
                na_rep="#NULL",
            )

            if options.verbose:
                util.log(f"Writing CSV to: {path}")

            self.client.hdfs.put(tmp_file_path, path)
        return path

    def get_schema(self):
        # define a temporary table using delimited data
        return sch.infer(self.df)

    def delimited_table(self, csv_dir, database=None):
        return self.client.delimited_file(
            csv_dir,
            self.get_schema(),
            name=f"ibis_tmp_pandas_{util.guid()}",
            database=database,
            delimiter=",",
            na_rep="#NULL",
            escapechar="\\\\",
            external=True,
            persist=False,
        )

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        with contextlib.suppress(com.IbisError):
            self.cleanup()
        return False

    def cleanup(self):
        while self.temp_hdfs_dirs:
            self.client.hdfs.rm(self.temp_hdfs_dirs.pop(), recursive=True)
