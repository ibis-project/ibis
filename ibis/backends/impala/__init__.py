"""Impala backend"""
import ibis.config
from ibis.backends.base import BaseBackend

# these objects are exposed in the public API and are not used in the module
from .client import (  # noqa: F401
    ImpalaClient,
    ImpalaConnection,
    ImpalaDatabase,
    ImpalaDatabaseTable,
    ImpalaTable,
)
from .hdfs import HDFS, WebHDFS, hdfs_connect
from .udf import *  # noqa: F401,F403


class Backend(BaseBackend):
    name = 'impala'
    kind = 'sql'
    client = ImpalaClient
    database_class = ImpalaDatabase
    table_class = ImpalaDatabaseTable
    table_expr_class = ImpalaTable
    HDFS = HDFS
    WebHDFS = WebHDFS

    def hdfs_connect(self, *args, **kwargs):
        return hdfs_connect(*args, **kwargs)

    def connect(
        self,
        host='localhost',
        port=21050,
        database='default',
        timeout=45,
        use_ssl=False,
        ca_cert=None,
        user=None,
        password=None,
        auth_mechanism='NOSASL',
        kerberos_service_name='impala',
        pool_size=8,
        hdfs_client=None,
    ):
        """Create an ImpalaClient for use with Ibis.

        Parameters
        ----------
        host : str, optional
            Host name of the impalad or HiveServer2 in Hive
        port : int, optional
            Impala's HiveServer2 port
        database : str, optional
            Default database when obtaining new cursors
        timeout : int, optional
            Connection timeout in seconds when communicating with HiveServer2
        use_ssl : bool, optional
            Use SSL when connecting to HiveServer2
        ca_cert : str, optional
            Local path to 3rd party CA certificate or copy of server
            certificate for self-signed certificates. If SSL is enabled, but
            this argument is ``None``, then certificate validation is skipped.
        user : str, optional
            LDAP user to authenticate
        password : str, optional
            LDAP password to authenticate
        auth_mechanism : str, optional
            {'NOSASL' <- default, 'PLAIN', 'GSSAPI', 'LDAP'}.
            Use NOSASL for non-secured Impala connections.  Use PLAIN for
            non-secured Hive clusters.  Use LDAP for LDAP authenticated
            connections.  Use GSSAPI for Kerberos-secured clusters.
        kerberos_service_name : str, optional
            Specify particular impalad service principal.

        Examples
        --------
        >>> import ibis
        >>> import os
        >>> hdfs_host = os.environ.get('IBIS_TEST_NN_HOST', 'localhost')
        >>> hdfs_port = int(os.environ.get('IBIS_TEST_NN_PORT', 50070))
        >>> impala_host = os.environ.get('IBIS_TEST_IMPALA_HOST', 'localhost')
        >>> impala_port = int(os.environ.get('IBIS_TEST_IMPALA_PORT', 21050))
        >>> hdfs = ibis.impala.hdfs_connect(host=hdfs_host, port=hdfs_port)
        >>> hdfs  # doctest: +ELLIPSIS
        <ibis.filesystems.WebHDFS object at 0x...>
        >>> client = ibis.impala.connect(
        ...     host=impala_host,
        ...     port=impala_port,
        ...     hdfs_client=hdfs,
        ... )
        >>> client  # doctest: +ELLIPSIS
        <ibis.impala.client.ImpalaClient object at 0x...>

        Returns
        -------
        ImpalaClient
        """
        params = {
            'host': host,
            'port': port,
            'database': database,
            'timeout': timeout,
            'use_ssl': use_ssl,
            'ca_cert': ca_cert,
            'user': user,
            'password': password,
            'auth_mechanism': auth_mechanism,
            'kerberos_service_name': kerberos_service_name,
        }

        con = ImpalaConnection(pool_size=pool_size, **params)
        try:
            client = ImpalaClient(
                backend=self, con=con, hdfs_client=hdfs_client
            )
        except Exception:
            con.close()
            raise
        return client

    def register_options(self):
        ibis.config.register_option(
            'temp_db',
            '__ibis_tmp',
            'Database to use for temporary tables, views. functions, etc.',
        )
        ibis.config.register_option(
            'temp_hdfs_path',
            '/tmp/ibis',
            'HDFS path for storage of temporary data',
        )
