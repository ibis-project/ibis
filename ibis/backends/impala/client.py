from __future__ import annotations

import contextlib
import traceback
from typing import TYPE_CHECKING

import sqlalchemy as sa
import sqlglot as sg

import ibis.common.exceptions as com
import ibis.expr.schema as sch
import ibis.expr.types as ir
from ibis import util
from ibis.backends.base import Database
from ibis.backends.base.sql.compiler import DDL, DML
from ibis.backends.base.sql.ddl import AlterTable, InsertSelect
from ibis.backends.impala import ddl
from ibis.backends.impala.compat import HS2Error, impyla

if TYPE_CHECKING:
    import pandas as pd


class ImpalaDatabase(Database):
    def create_table(self, name: str, obj=None, **kwargs) -> ir.Table:
        """Dispatch to ImpalaClient.create_table.

        See that function's docstring for more
        """
        return self.client.create_table(name, obj=obj, database=self.name, **kwargs)

    def list_udfs(self, like=None):
        return self.client.list_udfs(like=like, database=self.name)

    def list_udas(self, like=None):
        return self.client.list_udas(like=like, database=self.name)


class ImpalaConnection:
    """Database connection wrapper."""

    def __init__(self, pool_size=8, database="default", **params):
        self.params = params
        self.database = database
        self.options = {}
        self.pool = sa.pool.QueuePool(
            self._new_cursor,
            pool_size=pool_size,
            # disable invoking rollback, because any transactions in impala are
            # automatic:
            # https://impala.apache.org/docs/build/html/topics/impala_transactions.html
            reset_on_return=False,
        )

        @sa.event.listens_for(self.pool, "checkout")
        def _(dbapi_connection, *_):
            """Update `dbapi_connection` options if they don't match `self`."""
            if (options := self.options) != dbapi_connection.options:
                dbapi_connection.options = options.copy()
                dbapi_connection.set_options()

    def set_options(self, options):
        self.options.update(options)

    def close(self):
        """Close all idle Impyla connections."""
        self.pool.dispose()

    def disable_codegen(self, disabled=True):
        self.options["DISABLE_CODEGEN"] = str(int(disabled))

    def execute(self, query):
        if isinstance(query, (DDL, DML)):
            query = query.compile()

        util.log(query)

        cursor = self.pool.connect()
        try:
            cursor.execute(query)
        except Exception:
            cursor.close()
            util.log(f"Exception caused by {query}: {traceback.format_exc()}")
            raise

        return cursor

    def fetchall(self, query):
        cur = self.execute(query)
        try:
            results = cur.fetchall()
        finally:
            cur.close()
        return results

    def _new_cursor(self):
        params = self.params.copy()
        con = impyla.connect(database=self.database, **params)

        # make sure the connection works
        cursor = con.cursor(user=params.get("user"), convert_types=True)
        cursor.ping()

        wrapper = ImpalaCursor(cursor, con, self.database, self.options.copy())
        wrapper.set_options()
        return wrapper

    def release(self, cur):  # pragma: no cover
        pass


class ImpalaCursor:
    def __init__(self, cursor, impyla_con, database, options):
        self._cursor = cursor
        self.impyla_con = impyla_con
        self.database = database
        self.options = options

    def __del__(self):
        with contextlib.suppress(Exception):
            self.close()

    def close(self):
        try:
            self._cursor.close()
        except HS2Error as e:
            # connection was closed elsewhere
            already_closed_messages = [
                "invalid query handle",
                "invalid session",
            ]
            for message in already_closed_messages:
                if message in e.args[0].lower():
                    break
            else:
                raise

    def set_options(self):
        for k, v in self.options.items():
            query = f"SET {k} = {v!r}"
            self._cursor.execute(query)

    @property
    def description(self):
        return self._cursor.description

    def release(self):
        pass

    def execute(self, stmt):
        self._cursor.execute_async(stmt)
        self._wait_synchronous()

    def _wait_synchronous(self):
        # Wait to finish, but cancel if KeyboardInterrupt
        try:
            self._cursor._wait_to_finish()
        except KeyboardInterrupt:
            util.log("Canceling query")
            self.cancel()
            raise

    def is_finished(self):
        return not self.is_executing()

    def is_executing(self):
        return self._cursor.is_executing()

    def cancel(self):
        self._cursor.cancel_operation()

    def fetchone(self):
        return self._cursor.fetchone()

    def fetchall(self, columnar=False):
        if columnar:
            return self._cursor.fetchcolumnar()
        else:
            return self._cursor.fetchall()


class ImpalaTable(ir.Table):
    """A physical table in the Impala-Hive metastore."""

    @property
    def _qualified_name(self) -> str:
        op = self.op()
        return sg.table(op.name, db=op.namespace).sql(dialect="hive")

    @property
    def _unqualified_name(self) -> str:
        return self.op().name

    @property
    def _client(self):
        return self.op().source

    @property
    def _database(self) -> str:
        return self.op().namespace

    def compute_stats(self, incremental=False):
        """Invoke Impala COMPUTE STATS command on the table."""
        return self._client.compute_stats(self._qualified_name, incremental=incremental)

    def invalidate_metadata(self):
        self._client.invalidate_metadata(self._qualified_name)

    def refresh(self):
        self._client.refresh(self._qualified_name)

    def metadata(self):
        """Return results of `DESCRIBE FORMATTED` statement."""
        return self._client.describe_formatted(self._qualified_name)

    describe_formatted = metadata

    def files(self):
        """Return results of SHOW FILES statement."""
        return self._client.show_files(self._qualified_name)

    def drop(self):
        """Drop the table from the database."""
        self._client.drop_table_or_view(self._qualified_name)

    def truncate(self):
        self._client.truncate_table(self._qualified_name)

    def insert(
        self,
        obj=None,
        overwrite=False,
        partition=None,
        values=None,
        validate=True,
    ):
        """Insert into an Impala table.

        Parameters
        ----------
        obj
            Table expression or DataFrame
        overwrite
            If True, will replace existing contents of table
        partition
            For partitioned tables, indicate the partition that's being
            inserted into, either with an ordered list of partition keys or a
            dict of partition field name to value. For example for the
            partition (year=2007, month=7), this can be either (2007, 7) or
            {'year': 2007, 'month': 7}.
        values
            Unsupported and unused
        validate
            If True, do more rigorous validation that schema of table being
            inserted is compatible with the existing table

        Examples
        --------
        Append to an existing table

        >>> t.insert(table_expr)  # quartodoc: +SKIP # doctest: +SKIP

        Completely overwrite contents

        >>> t.insert(table_expr, overwrite=True)  # quartodoc: +SKIP # doctest: +SKIP
        """
        if values is not None:
            raise NotImplementedError

        with self._client._setup_insert(obj) as expr:
            if validate:
                existing_schema = self.schema()
                insert_schema = expr.schema()
                if not insert_schema.equals(existing_schema):
                    _validate_compatible(insert_schema, existing_schema)

            if partition is not None:
                partition_schema = self.partition_schema()
                partition_schema_names = frozenset(partition_schema.names)
                expr = expr.select(
                    [
                        column
                        for column in expr.columns
                        if column not in partition_schema_names
                    ]
                )
            else:
                partition_schema = None

            ast = self._client.compiler.to_ast(expr)
            select = ast.queries[0]
            statement = InsertSelect(
                self._qualified_name,
                select,
                partition=partition,
                partition_schema=partition_schema,
                overwrite=overwrite,
            )
            self._client._safe_exec_sql(statement.compile())
            return self

    def load_data(self, path, overwrite=False, partition=None):
        """Load data into an Impala table.

        Parameters
        ----------
        path
            Data to load
        overwrite
            Overwrite the existing data in the entire table or indicated
            partition
        partition
            If specified, the partition must already exist
        """
        if partition is not None:
            partition_schema = self.partition_schema()
        else:
            partition_schema = None

        stmt = ddl.LoadData(
            self._qualified_name,
            path,
            partition=partition,
            partition_schema=partition_schema,
            overwrite=overwrite,
        )

        self._client._safe_exec_sql(stmt.compile())
        return self

    @property
    def name(self) -> str:
        return self.op().name

    @property
    def is_partitioned(self):
        """True if the table is partitioned."""
        return self.metadata().is_partitioned

    def partition_schema(self):
        """Return the schema for the partition columns."""
        schema = self.schema()
        result = self.partitions()

        partition_fields = []
        for col in result.columns:
            if col not in schema:
                break
            partition_fields.append((col, schema[col]))

        return sch.Schema(dict(partition_fields))

    def add_partition(self, spec, location=None):
        """Add a new table partition.

        This API creates any necessary new directories in HDFS.

        Partition parameters can be set in a single DDL statement or you can
        use `alter_partition` to set them after the fact.
        """
        part_schema = self.partition_schema()
        stmt = ddl.AddPartition(
            self._qualified_name, spec, part_schema, location=location
        )
        self._client._safe_exec_sql(stmt)
        return self

    def alter(
        self,
        location=None,
        format=None,
        tbl_properties=None,
        serde_properties=None,
    ):
        """Change settings and parameters of the table.

        Parameters
        ----------
        location
            For partitioned tables, you may want the alter_partition function
        format
            Table format
        tbl_properties
            Table properties
        serde_properties
            Serialization/deserialization properties
        """

        def _run_ddl(**kwds):
            stmt = AlterTable(self._qualified_name, **kwds)
            self._client._safe_exec_sql(stmt)
            return self

        return self._alter_table_helper(
            _run_ddl,
            location=location,
            format=format,
            tbl_properties=tbl_properties,
            serde_properties=serde_properties,
        )

    def set_external(self, is_external=True):
        """Toggle the `EXTERNAL` table property."""
        self.alter(tbl_properties={"EXTERNAL": is_external})

    def alter_partition(
        self,
        spec,
        location=None,
        format=None,
        tbl_properties=None,
        serde_properties=None,
    ):
        """Change settings and parameters of an existing partition.

        Parameters
        ----------
        spec
            The partition keys for the partition being modified
        location
            Location of the partition
        format
            Table format
        tbl_properties
            Table properties
        serde_properties
            Serialization/deserialization properties
        """
        part_schema = self.partition_schema()

        def _run_ddl(**kwds):
            stmt = ddl.AlterPartition(self._qualified_name, spec, part_schema, **kwds)
            self._client._safe_exec_sql(stmt)
            return self

        return self._alter_table_helper(
            _run_ddl,
            location=location,
            format=format,
            tbl_properties=tbl_properties,
            serde_properties=serde_properties,
        )

    def _alter_table_helper(self, f, **alterations):
        results = []
        for k, v in alterations.items():
            if v is None:
                continue
            result = f(**{k: v})
            results.append(result)
        return results

    def drop_partition(self, spec):
        """Drop an existing table partition."""
        part_schema = self.partition_schema()
        stmt = ddl.DropPartition(self._qualified_name, spec, part_schema)
        self._client._safe_exec_sql(stmt)
        return self

    def partitions(self):
        """Return information about the table's partitions.

        Raises an exception if the table is not partitioned.
        """
        return self._client.list_partitions(self._qualified_name)

    def stats(self) -> pd.DataFrame:
        """Return results of `SHOW TABLE STATS`.

        If not partitioned, contains only one row.

        Returns
        -------
        DataFrame
            Table statistics
        """
        return self._client.table_stats(self._qualified_name)

    def column_stats(self) -> pd.DataFrame:
        """Return results of `SHOW COLUMN STATS`.

        Returns
        -------
        DataFrame
            Column statistics
        """
        return self._client.column_stats(self._qualified_name)


# ----------------------------------------------------------------------
# ORM-ish usability layer


class ScalarFunction:
    def drop(self):
        pass


class AggregateFunction:
    def drop(self):
        pass


def _validate_compatible(from_schema, to_schema):
    if set(from_schema.names) != set(to_schema.names):
        raise com.IbisInputError("Schemas have different names")

    for name in from_schema:
        lt = from_schema[name]
        rt = to_schema[name]
        if not lt.castable(rt):
            raise com.IbisInputError(f"Cannot safely cast {lt!r} to {rt!r}")
