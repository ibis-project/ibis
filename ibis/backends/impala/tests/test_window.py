from __future__ import annotations

import pytest
from pytest import param

import ibis
import ibis.common.exceptions as com
from ibis import window
from ibis.backends.impala.compiler import ImpalaCompiler
from ibis.tests.util import assert_equal

pytest.importorskip("impala")


@pytest.fixture(scope="module")
def alltypes(mockcon):
    return mockcon.table("alltypes")


def assert_sql_equal(expr, snapshot, out="out.sql"):
    result = ImpalaCompiler.to_sql(expr)
    snapshot.assert_match(result, out)


def test_aggregate_in_projection(alltypes, snapshot):
    t = alltypes
    proj = t[t, (t.f / t.f.sum()).name("normed_f")]
    assert_sql_equal(proj, snapshot)


def test_add_default_order_by(alltypes, snapshot):
    t = alltypes

    first = t.f.first().name("first")
    last = t.f.last().name("last")
    lag = t.f.lag().name("lag")
    diff = (t.f.lead() - t.f).name("fwd_diff")
    lag2 = t.f.lag().over(window(order_by=t.d)).name("lag2")
    grouped = t.group_by("g")
    proj = grouped.mutate([lag, diff, first, last, lag2])
    assert_sql_equal(proj, snapshot)


@pytest.mark.parametrize(
    "window",
    [
        param(window(preceding=0), id="prec_0"),
        param(window(following=0), id="foll_0"),
        param(window(preceding=5), id="prec_5"),
        param(window(preceding=5, following=0), id="prec_5_foll_0"),
        param(window(preceding=5, following=2), id="prec_5_foll_2"),
        param(window(following=2), id="foll_2"),
        param(window(following=2, preceding=0), id="foll_2_prec_0"),
        param(window(following=(5, 10)), id="foll_5_10"),
        param(window(preceding=(10, 5)), id="foll_10_5"),
        param(ibis.cumulative_window(), id="cumulative"),
        param(ibis.trailing_window(10), id="trailing_10"),
    ],
)
def test_window_frame_specs(alltypes, window, snapshot):
    t = alltypes

    w2 = window.order_by(t.f)
    expr = t.select(foo=t.d.sum().over(w2))
    assert_sql_equal(expr, snapshot)


def test_window_rows_with_max_lookback(alltypes):
    mlb = ibis.rows_with_max_lookback(3, ibis.interval(days=3))
    t = alltypes
    w = ibis.trailing_window(mlb, order_by=t.i)
    expr = t.a.sum().over(w)
    with pytest.raises(NotImplementedError):
        ImpalaCompiler.to_sql(expr)


@pytest.mark.parametrize("name", ["sum", "min", "max", "mean"])
def test_cumulative_functions(alltypes, name, snapshot):
    t = alltypes
    w = ibis.window(order_by=t.d)

    func = getattr(t.f, name)
    cumfunc = getattr(t.f, f"cum{name}")

    expr = cumfunc().over(w).name("foo")
    expected = func().over(ibis.cumulative_window(order_by=t.d)).name("foo")

    expr1 = t.select(expr)
    expr2 = t.select(expected)

    assert_sql_equal(expr1, snapshot, "out1.sql")
    assert_sql_equal(expr2, snapshot, "out2.sql")


def test_nested_analytic_function(alltypes, snapshot):
    t = alltypes

    w = window(order_by=t.f)
    expr = (t.f - t.f.lag()).lag().over(w).name("foo")
    result = t.select(expr)
    assert_sql_equal(result, snapshot)


def test_rank_functions(alltypes, snapshot):
    t = alltypes

    proj = t[t.g, t.f.rank().name("minr"), t.f.dense_rank().name("denser")]
    assert_sql_equal(proj, snapshot)


def test_multiple_windows(alltypes, snapshot):
    t = alltypes

    w = window(group_by=t.g)

    expr = t.f.sum().over(w) - t.f.sum()
    proj = t.select(t.g, result=expr)

    assert_sql_equal(proj, snapshot)


def test_order_by_desc(alltypes, snapshot):
    t = alltypes

    w = window(order_by=ibis.desc(t.f))

    proj = t[t.f, ibis.row_number().over(w).name("revrank")]
    assert_sql_equal(proj, snapshot, "out1.sql")

    expr = t.group_by("g").order_by(ibis.desc(t.f))[t.d.lag().name("foo"), t.a.max()]
    assert_sql_equal(expr, snapshot, "out2.sql")


def test_row_number_does_not_require_order_by(alltypes, snapshot):
    t = alltypes

    expr = t.group_by(t.g).mutate(ibis.row_number().name("foo"))
    assert_sql_equal(expr, snapshot, "out1.sql")

    expr = t.group_by(t.g).order_by(t.f).mutate(ibis.row_number().name("foo"))
    assert_sql_equal(expr, snapshot, "out2.sql")


def test_row_number_properly_composes_with_arithmetic(alltypes, snapshot):
    t = alltypes
    w = ibis.window(order_by=t.f)
    expr = t.mutate(new=ibis.row_number().over(w) / 2)
    assert_sql_equal(expr, snapshot)


@pytest.mark.parametrize(
    ["column", "op"],
    [("f", "approx_nunique"), ("f", "approx_median"), ("g", "group_concat")],
)
def test_unsupported_aggregate_functions(alltypes, column, op):
    t = alltypes
    w = ibis.window(order_by=t.d)
    expr = getattr(t[column], op)()
    proj = t.select(foo=expr.over(w))
    with pytest.raises(com.TranslationError):
        ImpalaCompiler.to_sql(proj)


def test_propagate_nested_windows(alltypes, snapshot):
    # GH #469
    t = alltypes

    w = ibis.window(group_by=t.g, order_by=t.f)

    col = (t.f - t.f.lag()).lag()

    # propagate down here!
    result = col.over(w)
    ex_expr = (t.f - t.f.lag().over(w)).lag().over(w)
    assert_equal(result, ex_expr)

    expr = t.select(col.over(w).name("foo"))
    assert_sql_equal(expr, snapshot)
