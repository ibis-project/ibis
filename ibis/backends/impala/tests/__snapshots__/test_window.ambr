# name: test_add_default_order_by
  '''
  SELECT *, lag(`f`) OVER (PARTITION BY `g` ORDER BY `f`) AS `lag`,
         lead(`f`) OVER (PARTITION BY `g` ORDER BY `f`) - `f` AS `fwd_diff`,
         first_value(`f`) OVER (PARTITION BY `g` ORDER BY `f`) AS `first`,
         last_value(`f`) OVER (PARTITION BY `g` ORDER BY `f`) AS `last`,
         lag(`f`) OVER (PARTITION BY `g` ORDER BY `d` ASC) AS `lag2`
  FROM alltypes
  '''
# ---
# name: test_aggregate_in_projection
  '''
  SELECT *, `f` / sum(`f`) OVER () AS `normed_f`
  FROM alltypes
  '''
# ---
# name: test_cumulative_functions[max]
  '''
  SELECT max(`f`) OVER (ORDER BY `d` ASC RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS `foo`
  FROM alltypes
  '''
# ---
# name: test_cumulative_functions[max].1
  '''
  SELECT max(`f`) OVER (ORDER BY `d` ASC RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS `foo`
  FROM alltypes
  '''
# ---
# name: test_cumulative_functions[mean]
  '''
  SELECT avg(`f`) OVER (ORDER BY `d` ASC RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS `foo`
  FROM alltypes
  '''
# ---
# name: test_cumulative_functions[mean].1
  '''
  SELECT avg(`f`) OVER (ORDER BY `d` ASC RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS `foo`
  FROM alltypes
  '''
# ---
# name: test_cumulative_functions[min]
  '''
  SELECT min(`f`) OVER (ORDER BY `d` ASC RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS `foo`
  FROM alltypes
  '''
# ---
# name: test_cumulative_functions[min].1
  '''
  SELECT min(`f`) OVER (ORDER BY `d` ASC RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS `foo`
  FROM alltypes
  '''
# ---
# name: test_cumulative_functions[sum]
  '''
  SELECT sum(`f`) OVER (ORDER BY `d` ASC RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS `foo`
  FROM alltypes
  '''
# ---
# name: test_cumulative_functions[sum].1
  '''
  SELECT sum(`f`) OVER (ORDER BY `d` ASC RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS `foo`
  FROM alltypes
  '''
# ---
# name: test_multiple_windows
  '''
  SELECT `g`, sum(`f`) OVER (PARTITION BY `g`) - sum(`f`) OVER () AS `result`
  FROM alltypes
  '''
# ---
# name: test_nested_analytic_function
  '''
  SELECT lag(`f` - lag(`f`) OVER (ORDER BY `f` ASC)) OVER (ORDER BY `f` ASC) AS `foo`
  FROM alltypes
  '''
# ---
# name: test_order_by_desc
  '''
  SELECT `f`, (row_number() OVER (ORDER BY `f` DESC) - 1) AS `revrank`
  FROM alltypes
  '''
# ---
# name: test_order_by_desc.1
  '''
  SELECT lag(`d`) OVER (PARTITION BY `g` ORDER BY `f` DESC) AS `foo`,
         max(`a`) OVER (PARTITION BY `g` ORDER BY `f` DESC) AS `max`
  FROM alltypes
  '''
# ---
# name: test_propagate_nested_windows
  '''
  SELECT lag(`f` - lag(`f`) OVER (PARTITION BY `g` ORDER BY `f` ASC)) OVER (PARTITION BY `g` ORDER BY `f` ASC) AS `foo`
  FROM alltypes
  '''
# ---
# name: test_rank_functions
  '''
  SELECT `g`, (rank() OVER (ORDER BY `f`) - 1) AS `minr`,
         (dense_rank() OVER (ORDER BY `f`) - 1) AS `denser`
  FROM alltypes
  '''
# ---
# name: test_row_number_does_not_require_order_by
  '''
  SELECT *, (row_number() OVER (PARTITION BY `g`) - 1) AS `foo`
  FROM alltypes
  '''
# ---
# name: test_row_number_does_not_require_order_by.1
  '''
  SELECT *, (row_number() OVER (PARTITION BY `g` ORDER BY `f`) - 1) AS `foo`
  FROM alltypes
  '''
# ---
# name: test_row_number_properly_composes_with_arithmetic
  '''
  SELECT *, (row_number() OVER (ORDER BY `f`) - 1) / 2 AS `new`
  FROM alltypes
  '''
# ---
# name: test_window_frame_specs[cumulative]
  '''
  SELECT sum(`d`) OVER (ORDER BY `f` ASC RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS `foo`
  FROM alltypes
  '''
# ---
# name: test_window_frame_specs[foll_0]
  '''
  SELECT sum(`d`) OVER (ORDER BY `f` ASC RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS `foo`
  FROM alltypes
  '''
# ---
# name: test_window_frame_specs[foll_10_5]
  '''
  SELECT sum(`d`) OVER (ORDER BY `f` ASC ROWS BETWEEN 10 PRECEDING AND 5 PRECEDING) AS `foo`
  FROM alltypes
  '''
# ---
# name: test_window_frame_specs[foll_2]
  '''
  SELECT sum(`d`) OVER (ORDER BY `f` ASC ROWS BETWEEN UNBOUNDED PRECEDING AND 2 FOLLOWING) AS `foo`
  FROM alltypes
  '''
# ---
# name: test_window_frame_specs[foll_2_prec_0]
  '''
  SELECT sum(`d`) OVER (ORDER BY `f` ASC ROWS BETWEEN CURRENT ROW AND 2 FOLLOWING) AS `foo`
  FROM alltypes
  '''
# ---
# name: test_window_frame_specs[foll_5_10]
  '''
  SELECT sum(`d`) OVER (ORDER BY `f` ASC ROWS BETWEEN 5 FOLLOWING AND 10 FOLLOWING) AS `foo`
  FROM alltypes
  '''
# ---
# name: test_window_frame_specs[prec_0]
  '''
  SELECT sum(`d`) OVER (ORDER BY `f` ASC RANGE BETWEEN CURRENT ROW AND UNBOUNDED FOLLOWING) AS `foo`
  FROM alltypes
  '''
# ---
# name: test_window_frame_specs[prec_5]
  '''
  SELECT sum(`d`) OVER (ORDER BY `f` ASC ROWS BETWEEN 5 PRECEDING AND UNBOUNDED FOLLOWING) AS `foo`
  FROM alltypes
  '''
# ---
# name: test_window_frame_specs[prec_5_foll_0]
  '''
  SELECT sum(`d`) OVER (ORDER BY `f` ASC ROWS BETWEEN 5 PRECEDING AND CURRENT ROW) AS `foo`
  FROM alltypes
  '''
# ---
# name: test_window_frame_specs[prec_5_foll_2]
  '''
  SELECT sum(`d`) OVER (ORDER BY `f` ASC ROWS BETWEEN 5 PRECEDING AND 2 FOLLOWING) AS `foo`
  FROM alltypes
  '''
# ---
# name: test_window_frame_specs[trailing_10]
  '''
  SELECT sum(`d`) OVER (ORDER BY `f` ASC ROWS BETWEEN 10 PRECEDING AND CURRENT ROW) AS `foo`
  FROM alltypes
  '''
# ---
